From f3ef68b945eebdd4b02b6625d7a29785bb13651f Mon Sep 17 00:00:00 2001
From: CIQ Kernel Automation <ciq_kernel_automation@ciq.com>
Date: Wed, 24 Jun 2026 07:45:39 +0000
Subject: [PATCH] wifi: mac80211: use safe list iteration in radar detect work

jira VULN-188783
cve CVE-2026-46166
commit-author Benjamin Berg <benjamin.berg@intel.com>
commit ac8eb3e18f41e2cc8492cc1d358bcb786c850270

The call to ieee80211_dfs_cac_cancel can cause the iterated chanctx to
be freed and removed from the list. Guard against this to avoid a
slab-use-after-free error.

	Cc: stable@vger.kernel.org
Fixes: bca8bc0399ac ("wifi: mac80211: handle ieee80211_radar_detected() for MLO")
	Signed-off-by: Benjamin Berg <benjamin.berg@intel.com>
Link: https://patch.msgid.link/20260505151539.236d63a1b736.I35dbb9e96a2d4a480be208770fdd99ba3b817b79@changeid
	Signed-off-by: Johannes Berg <johannes.berg@intel.com>
(cherry picked from commit ac8eb3e18f41e2cc8492cc1d358bcb786c850270)
	Signed-off-by: CIQ Kernel Automation <ciq_kernel_automation@ciq.com>
---
 net/mac80211/util.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/net/mac80211/util.c b/net/mac80211/util.c
index 38c30e4ddda9..3e474d0aed03 100644
--- a/net/mac80211/util.c
+++ b/net/mac80211/util.c
@@ -3501,11 +3501,11 @@ void ieee80211_dfs_radar_detected_work(struct wiphy *wiphy,
 	struct ieee80211_local *local =
 		container_of(work, struct ieee80211_local, radar_detected_work);
 	struct cfg80211_chan_def chandef;
-	struct ieee80211_chanctx *ctx;
+	struct ieee80211_chanctx *ctx, *tmp;
 
 	lockdep_assert_wiphy(local->hw.wiphy);
 
-	list_for_each_entry(ctx, &local->chanctx_list, list) {
+	list_for_each_entry_safe(ctx, tmp, &local->chanctx_list, list) {
 		if (ctx->replace_state == IEEE80211_CHANCTX_REPLACES_OTHER)
 			continue;
 
-- 
2.51.0

